parser_library/src/workspaces/macro_cache.cpp

```diff
@@ -0,0 +1,202 @@
+{
+    return { hlasm_ctx.opencode_file_name(), data, get_opsyn_state(hlasm_ctx) };
+}
+
+void macro_cache_key::sort_opsyn_state(std::vector<cached_opsyn_mnemo>& opsyn_state)
+{
+    std::sort(opsyn_state.begin(), opsyn_state.end(), [](const cached_opsyn_mnemo& lhs, const cached_opsyn_mnemo& rhs) {
+        return std::tie(lhs.from_instr, lhs.to_instr, lhs.is_macro)
+            < std::tie(rhs.from_instr, rhs.to_instr, rhs.is_macro);
+    });
+}
+
+const analyzer* macro_cache::find_cached_analyzer(const macro_cache_key& key) const
+{
+    auto it = cache_.find(key);
+    if (it == cache_.end())
+        return nullptr;
+
+    const auto& cached_data = it->second;
+
+    for (const auto& [fname, cached_version] : cached_data.stamps)
+    {
+        auto file = file_mngr_->find(fname);
+        if (!file)
+            return nullptr; // Reparse needed
+        if (file->get_version() != cached_version)
+            return nullptr;
+    }
+
+    // Version of all dependent files are the same.
+    return cached_data.cached_analyzer.get();
+}
+
+bool macro_cache::load_from_cache(const macro_cache_key& key, const analyzing_context& ctx)
+{
+    if (auto cached_analyzer = find_cached_analyzer(key))
+    {
+        if (key.data.proc_kind == processing::processing_kind::MACRO)
+        {
+            lsp::macro_info_ptr info = cached_analyzer->context().lsp_ctx->get_macro_info(key.data.library_member);
+            if (!info)
+                return true; // The file for which the analyzer is cached does not contain definition of macro
+            ctx.hlasm_ctx->add_macro(info->macro_definition);
+            ctx.lsp_ctx->add_macro(info, lsp::text_data_ref_t(macro_file_->get_text()));
+
+            // Add all copy members on which this macro is dependant
+            for (const auto& copy_ptr : info->macro_definition->used_copy_members)
+            {
+                auto copy_file = file_mngr_->find(copy_ptr->definition_location.file);
+                ctx.hlasm_ctx->add_copy_member(copy_ptr);
+                ctx.lsp_ctx->add_copy(copy_ptr, lsp::text_data_ref_t(copy_file->get_text()));
+            }
+        }
+        else if (key.data.proc_kind == processing::processing_kind::COPY)
+        {
+            auto copy_member = cached_analyzer->context().hlasm_ctx->get_copy_member(key.data.library_member);
+            ctx.hlasm_ctx->add_copy_member(copy_member);
+            ctx.lsp_ctx->add_copy(copy_member, lsp::text_data_ref_t(macro_file_->get_text()));
+        }
+
+        return true;
+    }
+    return false;
+}
+
+version_stamp macro_cache::get_copy_member_versions(context::macro_def_ptr macro) const
+{
+    version_stamp result;
+
+    for (const auto& copy_ptr : macro->used_copy_members)
+    {
+        auto file = file_mngr_->find(copy_ptr->definition_location.file);
+        if (!file)
+            throw std::runtime_error("Dependencies of a macro must be open right after parsing the macro.");
```

Comment:
how does this happen?

Reply:
I don't see that it could happen.. Should I remove the condition?

Reply:
I am again thinking about a race condition or something like that...